pennylane/_qubit_device.py

```diff
@@ -29,100 +30,164 @@
 class QubitDevice(Device):
     """Abstract base class for PennyLane qubit devices.
 
+    The following abstract methods **must** be defined:
+
+    * :meth:`~.probability`: returns the probability or marginal probability from the
+      device after circuit execution. :meth:`~.marginal_prob` may be used here.
+
+    * :meth:`~.apply`: append circuit operations, compile the circuit (if applicable),
+      and perform the quantum computation.
+
+    Where relevant, devices that generate their own samples (such as hardware) should
+    overwrite the following methods:
+
+    * :meth:`~.generate_samples`: Generate samples from the device from the
+      exact or approximate probability distribution.
+
+    This device contains common utility methods for qubit-based devices. These
+    do not need to be overwritten. Utility methods include:
+
+    * :meth:`~.expval`, :meth:`~.var`, :meth:`~.sample`: return expectation values,
+      variances, and samples of observables after the circuit has been rotated
+      into the observable eigenbasis.
+
     Args:
-        wires (int): number of subsystems in the quantum state represented by the device.
-            Default 1 if not specified.
-        shots (int): Number of circuit evaluations/random samples used to estimate
-            expectation values of observables. Defaults to 1000 if not specified.
+        wires (int): number of subsystems in the quantum state represented by the device
+        shots (int): number of circuit evaluations/random samples used to estimate
+            expectation values of observables
+        analytic (bool): If ``True``, the device calculates probability, expectation values,
+            and variances analytically. If ``False``, a finite number of samples set by
+            the argument ``shots`` are used to estimate these quantities.
     """
-    #pylint: disable=too-many-public-methods
+
+    # pylint: disable=too-many-public-methods
+    _asarray = staticmethod(np.asarray)
```

Comment:
Just a question here, would this be defined just for convenience?

Reply:
This is defined to allow `expt.tensornet.tf` to work, as it allows that device to overwrite this with `tf.convert_to_tensor()`.

Reply:
Gotcha, nice!